lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp

```diff
@@ -139,285 +135,379 @@
-    if (exit_block == latch_exit) {
-      return false;
+  void DeleteAllNewValues() {
+    for (auto &pair : cached_phis) {
+      pair.second->dropAllReferences();
     }
-
-    for (Instruction &I : *exit_block) {
-      if (PHINode *phi = dyn_cast<PHINode>(&I)) {
-        // If there are values flowing out of the loop into the exit_block,
-        // if any of those values do not dominate the latch, they would need
-        // to be propagated to the latch, which we don't do right now.
-        //
-        if (Instruction *value = dyn_cast<Instruction>(phi->getIncomingValueForBlock(exiting_block))) {
-          if (!DT->dominates(value, latch)) {
-            return false;
-          }
-        }
-      }
-      else {
-        break;
-      }
+    for (auto &pair : cached_phis) {
+      pair.second->eraseFromParent();
     }
+    cached_phis.clear();
   }
 
-  BranchInst *exiting_br = cast<BranchInst>(exiting_block->getTerminator());
-  Value *exit_cond = exiting_br->getCondition();
+  BasicBlock *Run(std::vector<Value_Info> &exit_values, BasicBlock *exiting_block, BasicBlock *latch, DominatorTree *DT, Loop *L, LoopInfo *LI, std::vector<BasicBlock *> &blocks_with_side_effect) {
+    BasicBlock *ret = RunImpl(exit_values, exiting_block, latch, DT, L, LI, blocks_with_side_effect);
+    // If we failed, remove all the values we added.
+    if (!ret) {
+      DeleteAllNewValues();
+    }
+    return ret;
+  }
 
-  Value *exit_cond_dominates_latch = nullptr;
-  BasicBlock *new_exiting_block = nullptr;
-  SmallVector<std::pair<BasicBlock *, Value *>, 4> blocks_with_side_effect;
-  bool give_up = false;
-  std::unordered_map<BasicBlock *, PHINode *> cached_phis;
+  BasicBlock *RunImpl(std::vector<Value_Info> &exit_values, BasicBlock *exiting_block, BasicBlock *latch, DominatorTree *DT, Loop *L, LoopInfo *LI, std::vector<BasicBlock *> &blocks_with_side_effect) {
 
-  // Use a worklist to propagate the exit condition from within the block
-  {
-    Value *false_value = ConstantInt::getFalse(i1Ty);
-
-    struct Propagate_Data {
+    struct Edge {
+      BasicBlock *prev;
       BasicBlock *bb;
-      Value *exit_cond;
     };
 
-    std::unordered_set<BasicBlock *> seen;
-    SmallVector<Propagate_Data, 4> work_list;
-
-    work_list.push_back({ exiting_block, exit_cond, });
+    BasicBlock *new_exiting_block = nullptr;
+    SmallVector<Edge, 4> work_list;
+    work_list.push_back({ nullptr, exiting_block });
     seen.insert(exiting_block);
 
-    BasicBlock *latch = L->getLoopLatch();
-
     for (unsigned i = 0; i < work_list.size(); i++) {
-      Propagate_Data data = work_list[i];
+      auto &edge = work_list[i];
+      BasicBlock *prev = edge.prev;
+      BasicBlock *bb = edge.bb;
 
-      BasicBlock *bb = data.bb;
-
-      // Don't continue to propagate when we hit the latch
+      // Don't continue to propagate when we hit the latch or dominate it.
       if (bb == latch && DT->dominates(bb, latch)) {
```

Comment:
This call to `dominates` is redundant, it will always return `true` after `bb` has been proven to be the same as `latch`.
The comment and rest of the code seems to indicate that this should be `||` instead of `&&`? In this case the check for equality is redundant.